File: gateway/src/test/java/com/beat/gateway/refreshtoken/internal/RefreshTokenServiceTest.java
package com.beat.gateway.refreshtoken.internal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.Optional;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.beat.gateway.refreshtoken.internal.store.RefreshToken;
import com.beat.gateway.refreshtoken.internal.store.RefreshTokenRepository;

@ExtendWith(MockitoExtension.class)
class RefreshTokenServiceTest {

	@Mock
	private RefreshTokenRepository refreshTokenRepository;

	private RefreshTokenService refreshTokenService;

	@BeforeEach
	void setUp() {
		refreshTokenService = new RefreshTokenService(refreshTokenRepository);
	}

	@Test
	void saveRefreshTokenStoresRefreshTokenAggregate() {
		refreshTokenService.saveRefreshToken(1L, "refresh-token");

		verify(refreshTokenRepository).save(argThat(token ->
			token.getId().equals(1L) && token.getRefreshToken().equals("refresh-token")
		));
	}

	@Test
	void findMemberIdByRefreshTokenReturnsMemberIdWhenTokenExists() {
		when(refreshTokenRepository.findByRefreshToken("refresh-token"))
			.thenReturn(Optional.of(RefreshToken.of(1L, "refresh-token")));

		long memberId = refreshTokenService.findMemberIdByRefreshToken("refresh-token").orElseThrow();

		assertEquals(1L, memberId);
	}

	@Test
	void findMemberIdByRefreshTokenReturnsEmptyWhenTokenIsMissing() {
		when(refreshTokenRepository.findByRefreshToken("missing")).thenReturn(Optional.empty());

		assertFalse(refreshTokenService.findMemberIdByRefreshToken("missing").isPresent());
	}

	@Test
	void deleteRefreshTokenDeletesLoadedToken() {
		RefreshToken token = RefreshToken.of(1L, "refresh-token");
		when(refreshTokenRepository.findById(1L)).thenReturn(Optional.of(token));

		boolean deleted = refreshTokenService.deleteRefreshToken(1L);

		assertTrue(deleted);
		verify(refreshTokenRepository).delete(token);
	}

	@Test
	void deleteRefreshTokenIsIdempotentWhenTokenIsMissing() {
		when(refreshTokenRepository.findById(1L)).thenReturn(Optional.empty());

		boolean deleted = refreshTokenService.deleteRefreshToken(1L);

		assertFalse(deleted);
		verify(refreshTokenRepository, never()).delete(any());
	}
}

File: gateway/src/test/kotlin/com/beat/gateway/refreshtoken/internal/RefreshTokenServiceTest.kt
package com.beat.gateway.refreshtoken.internal

import com.beat.gateway.refreshtoken.internal.store.RefreshToken
import com.beat.gateway.refreshtoken.internal.store.RefreshTokenRepository
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.argThat
import org.mockito.Mockito.mock
import org.mockito.Mockito.never
import org.mockito.Mockito.verify
import java.util.Optional
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.mockito.Mockito.`when` as given

class RefreshTokenServiceTest {

    private lateinit var refreshTokenRepository: RefreshTokenRepository
    private lateinit var refreshTokenService: RefreshTokenService

    @BeforeEach
    fun setUp() {
        refreshTokenRepository = mock(RefreshTokenRepository::class.java)
        refreshTokenService = RefreshTokenService(refreshTokenRepository)
    }

    @Test
    fun `refresh token을 memberId와 함께 저장한다`() {
        refreshTokenService.saveRefreshToken(1L, "refresh-token")

        verify(refreshTokenRepository).save(
            argThat { token: RefreshToken -> token.id == 1L && token.refreshToken == "refresh-token" },
        )
    }

    @Test
    fun `저장된 refresh token으로 memberId를 조회한다`() {
        given(refreshTokenRepository.findByRefreshToken("refresh-token"))
            .thenReturn(Optional.of(RefreshToken(1L, "refresh-token")))

        assertEquals(1L, refreshTokenService.findMemberIdByRefreshToken("refresh-token").orElseThrow())
    }

    @Test
    fun `존재하지 않는 refresh token 조회는 비어 있는 결과를 반환한다`() {
        given(refreshTokenRepository.findByRefreshToken("missing")).thenReturn(Optional.empty())

        assertFalse(refreshTokenService.findMemberIdByRefreshToken("missing").isPresent)
    }

    @Test
    fun `조회된 refresh token을 삭제한다`() {
        val token = RefreshToken(1L, "refresh-token")
        given(refreshTokenRepository.findById(1L)).thenReturn(Optional.of(token))

        assertTrue(refreshTokenService.deleteRefreshToken(1L))
        verify(refreshTokenRepository).delete(token)
    }

    @Test
    fun `삭제 대상이 없으면 멱등하게 false를 반환한다`() {
        given(refreshTokenRepository.findById(1L)).thenReturn(Optional.empty())

        assertFalse(refreshTokenService.deleteRefreshToken(1L))
        verify(refreshTokenRepository, never()).delete(any())
    }
}
